File: medium/2406_divide_intervals_into_minimum_number_of_groups/solution.cpp
#include <vector>

class Solution {
public:
  int minGroups(std::vector<std::vector<int>> &intervals) {
    std::sort(intervals.begin(), intervals.end());

    std::vector<int> rooms;

    for (const std::vector<int> &interval : intervals) {
      if (!rooms.empty() && -rooms[0] < interval[0]) {
        std::pop_heap(rooms.begin(), rooms.end());
        rooms.pop_back();
      }

      rooms.push_back(-interval[1]);
      std::push_heap(rooms.begin(), rooms.end());
    }

    return rooms.size();
  }
};

File: medium/2406_divide_intervals_into_minimum_number_of_groups/solution.py
import heapq
from typing import List


class Solution:
    def minGroups(self, intervals: List[List[int]]) -> int:
        rooms = []

        for start, end in sorted(intervals):
            if rooms and rooms[0] < start:
                heapq.heappop(rooms)
            heapq.heappush(rooms, end)

        return len(rooms)
